File: BFWControls/Modules/Blur/View/UIImage+BFW.m
//
//  UIImage+BFW.m
//
//  Created by Tom Brodhurst-Hill on 5/03/12.
//  Copyright (c) 2012 BareFeetWare. All rights reserved.
//  Free to use at your own risk, with acknowledgement to BareFeetWare.
//

#import "UIImage+BFW.h"
#import <QuartzCore/QuartzCore.h>

@implementation UIImage (BFW)

/// Adapted from http://iphonedevelopertips.com/cocoa/how-to-mask-an-image.html
- (UIImage*)maskWithImage:(UIImage*)maskImage
{
	CGImageRef maskRef = maskImage.CGImage; 
	CGImageRef mask = CGImageMaskCreate(CGImageGetWidth(maskRef),
										CGImageGetHeight(maskRef),
										CGImageGetBitsPerComponent(maskRef),
										CGImageGetBitsPerPixel(maskRef),
										CGImageGetBytesPerRow(maskRef),
										CGImageGetDataProvider(maskRef), NULL, false);
	
	CGImageRef masked = CGImageCreateWithMask([self CGImage], mask);
	return [UIImage imageWithCGImage:masked];
}

+ (UIImage*)imageOfView:(UIView*)view
                   size:(CGSize)size
{
	BOOL isOpaque = NO;
	CGFloat scale = [[UIScreen mainScreen] scale];
	UIGraphicsBeginImageContextWithOptions(size, isOpaque, scale);
	[view.layer renderInContext:UIGraphicsGetCurrentContext()];
	UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
	UIGraphicsEndImageContext();
	return image;
}

@end

File: BFWControls/Modules/Blur/View/UIImage+BFW.swift
//
//  UIImage+BFW.swift
//  BFWControls
//
//  Created by Charlotte Tortorella on 19/4/17.
//  Copyright © 2017 BareFeetWare. All rights reserved.
//

import Foundation

extension UIImage {
    func masking(_ image: UIImage) -> UIImage? {
        guard let maskRef = image.cgImage,
            let dataProvider = maskRef.dataProvider
            else { return nil }
        let mask = CGImage(maskWidth: maskRef.width,
                           height: maskRef.height,
                           bitsPerComponent: maskRef.bitsPerComponent,
                           bitsPerPixel: maskRef.bitsPerPixel,
                           bytesPerRow: maskRef.bytesPerRow,
                           provider: dataProvider,
                           decode: nil,
                           shouldInterpolate: false)
        return mask
            .flatMap { self.cgImage?.masking($0) }
            .flatMap(UIImage.init(cgImage:))
    }

    static func snapshot(of view: UIView, size: CGSize) -> UIImage? {
        UIGraphicsBeginImageContextWithOptions(size, false, UIScreen.main.scale)
        defer { UIGraphicsEndImageContext() }
        if let context = UIGraphicsGetCurrentContext() {
            view.layer.render(in: context)
        }
        return UIGraphicsGetImageFromCurrentImageContext()
    }
}
